### PR TITLE
Feature/iat 484

### PR DESCRIPTION
**Pull request by alex-ponce**
IAT-484
Initial requirement indicated to allow all characters and let the service indicate if the item id exists. 
The search term was URI encoded using encodeURIComponent(), however even after this it was discovered that several characters created invalid URI values.
The issue is caused because the search term is used to navigate to the /item/{itemId} route.
The solution implemented in this fix includes:
* URI encoding item id values
* Validate against the use of the following characters in the search term ~!@#$%^&*()
** If any of these characters is used an error message is displayed and the search does not take place

IAT-479
* Search terms with leading spaces no longer enable the Search button
* Search term is trimmed of leading and trailing spaces before page navigates to /item/{itemId}

### CODE REVIEW

File: src/app/item/components/find-item/find-item.component.html [hunk 1]
```diff
@@ -1,9 +1,14 @@
...
   <div class="col-md-8">
     <div class="input-group" [ngClass]="{'has-error':!complexForm.controls['query'].valid}">
-      <input type="text" class="form-control" placeholder="Item or stimulus ID"
-             [formControl]="complexForm.controls['query']" maxlength="16"
-             (keypress)="_keyPress($event)"/>
+      <input type="text"
+             class="form-control"
+             placeholder="Item or stimulus ID"
+             maxlength="16"
+             pattern="^(?!\s.*$).*"
```
Comment (bdrainer) [inline src/app/item/components/find-item/find-item.component.html]: very cool, I like the pattern on the html and not in the component

File: src/app/item/components/find-item/find-item.component.ts [hunk 2]
```diff
@@ -1,35 +1,42 @@
...
   }
 
   submitForm(value: any) {
-    this.logger.debug("Item search: " + value.query);
-    this.router.navigateByUrl('/item/' + value.query);
+    this.logger.debug("Item search: " + encodeURIComponent(value.query.trim()));
+    if (this.containsSpecialCharacters(value.query)) {
+      this.alertService.error("Invalid Search Term", "Your search term contains special characters. Please remove them and try again");
+    } else {
+      this.router.navigateByUrl('/item/' + encodeURIComponent(value.query.trim()));
```
Comment (bdrainer) [inline src/app/item/components/find-item/find-item.component.ts]: makes sense

File: src/app/item/services/item.service.ts [hunk 3]
```diff
@@ -38,7 +38,7 @@ export class ItemService {
   //---------------------------------------------------------------------------
   // Return the item with the given ID
   findItem(itemId: string): Observable<Item> {
-    const url = `${ItemService.serviceUrl}/${itemId}`;
+    const url = `${ItemService.serviceUrl}/${encodeURIComponent(itemId.trim())}`;
```
Comment (bdrainer) [inline src/app/item/services/item.service.ts]: This is the part I was was missing.  Nice job.

File: src/app/item/components/find-item/find-item.component.ts (hunk 2 above)
Comment (rmitchell3) [inline src/app/item/components/find-item/find-item.component.ts]: On line 30 please tell the user which characters are not allowed in the alert message.

File: src/app/item/components/find-item/find-item.component.html (hunk 1 above)
Comment (bdrainer) [inline src/app/item/components/find-item/find-item.component.html]: I tested this and wonder everyones thoughts... If a user enters invalid characters they get an error message.   If a user enters characters followed by spaces they are trimmed off and the form is submitted as expected.  However if the user first enters spaces and then characters the form is disabled with no indication of what is wrong.  What if we let them submit the form even with spaces in the front and then the trim will chop them off the front just like it does the back?

Comment (rmitchell3) [inline src/app/item/components/find-item/find-item.component.html]: I agree.  Spaces in the front, or back, or both should behave the same.

**Pull request merged**